contrib/IECoreUSD/src/IECoreUSD/ObjectAlgo.cpp

```diff
@@ -0,0 +1,125 @@
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "ObjectAlgo.h"
+
+//////////////////////////////////////////////////////////////////////////
+// Reading
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+struct ReaderFunctions
+{
+	IECoreUSD::ObjectAlgo::Reader reader;
+	IECoreUSD::ObjectAlgo::MightBeTimeVarying mightBeTimeVarying;
+};
+
+using ReaderMap = std::map<pxr::TfToken, ReaderFunctions>;
+ReaderMap &readers()
+{
+	static ReaderMap g_readers;
+	return g_readers;
+}
+
+} // namespace
+
+void IECoreUSD::ObjectAlgo::registerReader( const pxr::TfToken &schemaType, IECoreUSD::ObjectAlgo::Reader reader, IECoreUSD::ObjectAlgo::MightBeTimeVarying mightBeTimeVarying )
+{
+	readers()[schemaType] = { reader, mightBeTimeVarying };
+}
+
+bool IECoreUSD::ObjectAlgo::canReadObject( const pxr::UsdPrim &prim )
+{
+	const ReaderMap &r = readers();
+	return r.find( prim.GetTypeName() ) != r.end();
+}
+
+IECore::ObjectPtr IECoreUSD::ObjectAlgo::readObject( const pxr::UsdPrim &prim, pxr::UsdTimeCode time )
+{
+	const ReaderMap &r = readers();
+	auto it = r.find( prim.GetTypeName() );
+	if( it == r.end() )
+	{
+		return nullptr;
+	}
+
+	return it->second.reader( prim, time );
+}
+
+bool IECoreUSD::ObjectAlgo::objectMightBeTimeVarying( const pxr::UsdPrim &prim )
+{
+	const ReaderMap &r = readers();
+	auto it = r.find( prim.GetTypeName() );
+	if( it == r.end() )
+	{
+		return false;
+	}
+
+	return it->second.mightBeTimeVarying( prim );
+}
+
+//////////////////////////////////////////////////////////////////////////
+// Writing
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+using WriterMap = std::unordered_map<IECore::TypeId, IECoreUSD::ObjectAlgo::Writer>;
+
+WriterMap &writers()
+{
+	static WriterMap g_writers;
+	return g_writers;
+}
+
+} // namespace
+
+void IECoreUSD::ObjectAlgo::registerWriter( const IECore::TypeId typeId, IECoreUSD::ObjectAlgo::Writer writer )
+{
+	writers()[typeId] = writer;
+}
+
+void IECoreUSD::ObjectAlgo::writeObject( const IECore::Object *object, const pxr::UsdStagePtr &stage, const pxr::SdfPath &path, pxr::UsdTimeCode time )
+{
+	const auto &w = writers();
+	auto it = w.find( object->typeId() );
+	if( it == w.end() )
+	{
+		return;
```

Comment:
Should this throw or warn or return a bool success? A silent return would leave the client thinking the object was successfully written.

Reply:
In a1f1c91 I've gone with a bool return for `writeObject()`, thinking that this is more suitable than a message for a lower-level function, as it allows the caller to decide what to do. In the same commit I've turned this failure into a warning message from `USDScene::writeObject()`, thinking that an exception would be too harsh for a higher-level function (we can't use a return type without breaking ABI). This seems to be roughly in keeping with what we're doing in AlembicScene, and means Gaffer's SceneWriter won't fail to write completely.